internal/controller/nnf_access_controller.go

```diff
@@ -1149,6 +1161,137 @@ func clientMountName(access *nnfv1alpha7.NnfAccess) string {
+		if nnfAccess.GetDeletionTimestamp().Add(time.Duration(time.Duration(childTimeout) * time.Second)).After(time.Now()) {
+			return nil
+		}
+	}
+
+	// Find all ClientMounts owned by the NnfAccess
+	clientMountList := &dwsv1alpha3.ClientMountList{}
+	matchLabels := dwsv1alpha3.MatchingOwner(nnfAccess)
+
+	listOptions := []client.ListOption{
+		matchLabels,
+	}
+
+	if err := r.List(ctx, clientMountList, listOptions...); err != nil {
+		return dwsv1alpha3.NewResourceError("could not list ClientMounts").WithError(err)
+	}
+
+	// For each ClientMount, check whether the Compute node is marked as "Offline". If it is,
+	// remove the NnfClientMount finalizer and update the resource
+	for _, clientMount := range clientMountList.Items {
+		offline, err := r.checkOfflineCompute(ctx, nnfAccess, &clientMount)
+		if err != nil {
+			return err
+		}
+
+		if !offline {
+			continue
+		}
+
+		// Remove the finalizer from the ClientMount since the compute node is offline
+		controllerutil.RemoveFinalizer(&clientMount, finalizerNnfClientMount)
+		if err := r.Update(ctx, &clientMount); err != nil {
+			if !apierrors.IsConflict(err) {
+				return dwsv1alpha3.NewResourceError("could not update ClientMount to remove finalizer for offline compute: %v", client.ObjectKeyFromObject(&clientMount)).WithError(err)
+			}
+		}
+	}
+
+	return nil
+}
+
+// checkOfflineCompute finds the Storage resource for the Rabbit physically attached to the clientMount's
+// compute node, and checks whether the compute node is marked as "Offline" (indicating no PCIe connection)
+func (r *NnfAccessReconciler) checkOfflineCompute(ctx context.Context, nnfAccess *nnfv1alpha7.NnfAccess, clientMount *dwsv1alpha3.ClientMount) (bool, error) {
+	if nnfAccess.Spec.ClientReference == (corev1.ObjectReference{}) {
+		return false, nil
+	}
+
+	// Find the name of the Rabbit attached to the compute node
+	rabbitName, err := r.getRabbitFromClientMount(ctx, clientMount)
+	if err != nil {
+		return false, err
+	}
+
+	// Get the Storage resource for the Rabbit
+	storage := &dwsv1alpha3.Storage{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      rabbitName,
+			Namespace: "default",
+		},
+	}
+
+	if err := r.Get(ctx, client.ObjectKeyFromObject(storage), storage); err != nil {
+		return false, dwsv1alpha3.NewResourceError("could not get Storage %v", client.ObjectKeyFromObject(storage)).WithError(err).WithMajor()
+	}
+
+	// Check the status of the compute node in the Storage resource to determine if it's offline
+	computeName := clientMount.GetNamespace()
+	for _, compute := range storage.Status.Access.Computes {
+		if compute.Name == computeName {
+			return compute.Status == dwsv1alpha3.OfflineStatus, nil
+		}
+	}
+
+	return false, nil
+}
+
+// getRabbitFromClientMount finds the name of the Rabbit that is physically connected to the ClientMount's
+// compute node
+func (r *NnfAccessReconciler) getRabbitFromClientMount(ctx context.Context, clientMount *dwsv1alpha3.ClientMount) (string, error) {
```

Comment:
Similar to Tony's comment - and I believe we talked about this during standup, but could this be made available too? nnf-dm needs to do something similar and it would be nice to have an "official" way to do this.

Reply:
During standup we talked about putting it under sos's pkg/ dir.